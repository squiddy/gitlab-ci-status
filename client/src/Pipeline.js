import React from "react";

import { Avatar } from "./Avatar";
import { StatusIcon } from "./StatusIcon";

export function getTotalBuildRunTimeMs(builds) {
  return builds.reduce((sum, b) => {
    if (!b.started_at) {
      return sum;
    }

    const start = new Date(b.started_at.slice(0, -4));

    if (!b.finished_at) {
      return sum + (new Date() - start) / 1000;
    }

    const finish = new Date(b.finished_at.slice(0, -4));
    return sum + (finish - start) / 1000;
  }, 0);
}

export function Pipeline({ pipeline }) {
  const duration = getTotalBuildRunTimeMs(pipeline.builds);
  const isFinished = !["created", "pending", "running"].includes(
    pipeline.status
  );

  return (
    <div className={`queue-entry`}>
      <div className="queue-entry-body">
        <div className="avatars">
          <Avatar obj={pipeline.user} />
          <Avatar obj={pipeline.project} />
        </div>
        <div className="info">
          <strong>{pipeline.project.name}</strong>
          {pipeline.ref}
        </div>
        <div className={`status-info status-${pipeline.status}`}>
          {pipeline.status}
          <StatusIcon status={pipeline.status} />
        </div>
      </div>
      <div className="queue-entry-footer">
        <div className="pipeline-duration">
          {isFinished ? "took" : "running for"}
          {' '}
          {Math.floor(duration / 60)}m {duration % 60}s
        </div>
        <PipelineGraph builds={pipeline.builds} />
      </div>
    </div>
  );
}

export function PipelineGraph({ builds }) {
  return (
    <ul className="pipeline-graph">
      {builds.filter(b => b).map((b, idx) => (
        <li key={idx}>
          <StatusIcon title={b._raw.name} status={b.status} />
        </li>
      ))}
    </ul>
  );
}
