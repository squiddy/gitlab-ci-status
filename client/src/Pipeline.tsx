import React, { useState } from 'react';

import { Avatar } from './Avatar';
import { StatusIcon } from './StatusIcon';
import { Duration } from './Duration';
import { parseDate, isPipelineFinished } from './utils';
import { PipelineGraph } from './PipelineGraph';
import { BuildData, PipelineData, Status } from './types';

export function getTotalBuildRunTimeMs(builds: BuildData[]): number {
  return builds.reduce((sum, b) => {
    if (!b.started_at) {
      return sum;
    }

    const start = parseDate(b.started_at);

    if (!b.finished_at) {
      return sum + (new Date().getTime() - start.getTime()) / 1000;
    }

    const finish = parseDate(b.finished_at);
    return sum + (finish.getTime() - start.getTime()) / 1000;
  }, 0);
}

export function getRealDuration(pipeline: PipelineData): number | null {
  const data = pipeline._raw.object_attributes;
  if (!data.created_at || !data.finished_at) {
    return null;
  }

  return (
    parseDate(data.finished_at).getTime() / 1000 -
    parseDate(data.created_at).getTime() / 1000
  );
}

export function Pipeline({ pipeline }: { pipeline: PipelineData }) {
  const [showDetails, setShowDetails] = useState(
    !isPipelineFinished(pipeline.status) || pipeline.status === 'failed'
  );

  const duration = getTotalBuildRunTimeMs(pipeline.builds);
  const realDuration = getRealDuration(pipeline);

  const navigateToGitLab = () => {
    const url = `http://gitlab.bof.mm.local/${pipeline._raw.project.path_with_namespace}/pipelines/${pipeline.id}`;
    window.open(url, '_blank');
    window.focus();
  };

  const isMainRepository =
    pipeline._raw.project.namespace !== pipeline._raw.user.username;

  return (
    <div
      className={`flex flex-col mb-8 rounded cursor-pointer ${
        isMainRepository ? 'main-repository' : ''
      }`}
    >
      <div
        className="flex flex-row justify-between items-center bg-white rounded-t"
        onClick={navigateToGitLab}
      >
        <div className="p-4 flex-initial w-48">
          <Avatar className="h-16 rounded-full m-1" obj={pipeline.project} />
          <Avatar className="h-16 rounded-full m-1" obj={pipeline.user} />
        </div>
        <div className="flex-1 px-6 py-4">
          <div className="font-bold text-xl mb-1">{pipeline.project.name}</div>
          <p className="text-grey-darker text-base mb-2">{pipeline.ref}</p>
          <small className="text-grey-dark text-sm">
            {pipeline._raw.commit.message}
          </small>
        </div>
        <div className="p-4 flex-initial flex items-center">
          <StatusIcon className="h-12 w-12" status={pipeline.status} />
        </div>
      </div>
      <div
        className={'flex flex-col bg-indigo-darker rounded-b text-grey'}
        onClick={() => setShowDetails(!showDetails)}
      >
        {showDetails && (
          <div className="border-b border-dotted border-indigo-darkest px-4 py-4">
            <PipelineGraph pipeline={pipeline} />
          </div>
        )}
        <div className="flex justify-between items-center w-full px-4 py-2">
          <div className="text-xs">
            {realDuration ? (
              <>
                took <Duration value={realDuration} /> (
                <Duration value={duration} />)
              </>
            ) : (
              pipeline.status !== Status.Skipped &&
              pipeline.status !== Status.Canceled && (
                <>
                  running for <Duration value={duration} ticking />
                </>
              )
            )}
          </div>
        </div>
      </div>
    </div>
  );
}
