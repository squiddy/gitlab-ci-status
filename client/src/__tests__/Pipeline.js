import React from "react";
import renderer from "react-test-renderer";

import { PipelineGraph } from "../Pipeline";

it("PipelineGraph renders correctly", () => {
  const builds = [
    {
      stage: "check",
      status: "success"
    },
    {
      stage: "test",
      status: "running"
    }
  ];
  const tree = renderer.create(<PipelineGraph builds={builds} />).toJSON();
  expect(tree).toMatchSnapshot();
});
