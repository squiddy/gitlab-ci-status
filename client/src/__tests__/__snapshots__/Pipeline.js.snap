// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelineGraph renders correctly 1`] = `
<ul
  className="pipeline-graph"
>
  <li>
    <img
      alt=""
      src="success.svg"
      title="check"
    />
  </li>
  <li>
    <img
      alt=""
      src="running.svg"
      title="test"
    />
  </li>
</ul>
`;
